src/kernels/jupyter/session/jupyterSessionManager.ts

```diff
@@ -221,34 +226,103 @@ export class JupyterSessionManager implements IJupyterSessionManager {
-            await Promise.race([sleep(10_000), this.specsManager?.refreshSpecs()]);
-
-            // Enumerate all of the kernel specs, turning each into a JupyterKernelSpec
-            const kernelspecs =
-                this.specsManager?.specs && Object.keys(this.specsManager.specs.kernelspecs).length
-                    ? this.specsManager.specs.kernelspecs
-                    : oldKernelSpecs;
-            const keys = Object.keys(kernelspecs);
-            if (keys && keys.length) {
-                return keys.map((k) => {
-                    const spec = kernelspecs[k];
-                    return new JupyterKernelSpec(spec!) as IJupyterKernelSpec;
+            await Promise.race([sleep(10_000), specsManager.refreshSpecs()]);
+            telemetryProperties.specsManagerReady = specsManager.isReady;
+
+            let telemetrySent = false;
+            if (specsManager && Object.keys(specsManager.specs?.kernelspecs || {}).length === 0) {
+                // At this point wait for the specs to change
+                const disposables: IDisposable[] = [];
+                const promise = createDeferred();
+                specsManager.specsChanged.connect(promise.resolve);
+                disposables.push(new Disposable(() => specsManager.specsChanged.disconnect(promise.resolve)));
+                const allPromises = Promise.all([
+                    promise.promise,
+                    specsManager.ready,
+                    specsManager.refreshSpecs(),
+                    this.sessionManager.ready
+                ]);
+                await Promise.race([sleep(10_000), allPromises]);
+                telemetryProperties.waitedForChangeEvent = true;
+                if (!promise.completed) {
+                    telemetrySent = true;
+                    sendTelemetryEvent(Telemetry.JupyterKernelSpecEnumeration, undefined, {
+                        failed: true,
+                        sessionManagerReady: this.sessionManager.isReady,
+                        specsManagerReady: specsManager.isReady,
+                        reason: specsManager.isReady
+                            ? this.sessionManager.isReady
+                                ? 'SpecsDidNotChangeInTime'
+                                : 'SessionManagerIsNotReady'
+                            : 'SpecManagerIsNotReady'
+                    });
+                }
+                disposeAllDisposables(disposables);
+            }
+
+            const kernelspecs = getKernelSpecs(oldKernelSpecs);
+            if (Object.keys(kernelspecs || {}).length) {
+                const specs: IJupyterKernelSpec[] = [];
+                Object.entries(kernelspecs).forEach(([_key, value]) => {
+                    if (value) {
+                        specs.push(new JupyterKernelSpec(value));
+                    }
                 });
+                sendTelemetryEvent(
+                    Telemetry.JupyterKernelSpecEnumeration,
```

Comment:
Sending telemetry to see how often we end up having to wait for this event.
If we have to wait for this event 100% of the time then we might as well always do this.
also measuring the time to see if the timeout of 10s is required (the jupyter code seem to have some code that tries every 60s)